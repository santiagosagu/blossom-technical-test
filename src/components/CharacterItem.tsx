import { Heart, Trash } from "lucide-react";
import type { Character } from "../types/character";
import {
  useCharacter,
  type CharacterReturnType,
} from "../store/characterStore";
import { useNavigate } from "react-router";
import { ConfirmDeleteDialog } from "./ui/ConfirmDialog";
import { useState } from "react";

interface CharacterItemProps {
  character: Character;
  isSelected: boolean;
}

export default function CharacterItem({
  character,
  isSelected,
}: CharacterItemProps) {
  const { toggleFavorite, toggleDelete, setSelectedCharacter } =
    useCharacter() as CharacterReturnType;
  const navigate = useNavigate();

  const [open, setOpen] = useState(false);

  return (
    <div
      className={`flex items-center gap-3 p-3  cursor-pointer transition-colors border-t-gray-100 border-t-1 md:rounded-lg  ${
        isSelected ? "md:bg-primary-100 " : "hover:bg-gray-100 "
      }`}
      onClick={(e) => {
        e.stopPropagation();
        setSelectedCharacter(character);

        if (!open) {
          navigate(`/character-detail/${character.id}`);
        }
      }}
    >
      <div className="relative">
        <img
          src={character.image as string}
          alt={character.name as string}
          className="w-10 h-10 rounded-full object-cover"
        />
      </div>

      <div className="flex-1 min-w-0">
        <h3 className="font-medium text-sm truncate">{character.name}</h3>
        <p className="text-xs text-gray-500">{character.species}</p>
      </div>

      <button
        aria-label="favorite-button"
        onClick={(e) => {
          e.stopPropagation();
          toggleFavorite(character.id as string);
        }}
        className="p-1 h-auto cursor-pointer"
      >
        <Heart
          data-testid="heart-button"
          className={`w-4 h-4 ${
            character.isFavorite
              ? "fill-secondary-600 text-secondary-600"
              : "text-gray-400"
          }`}
        />
      </button>
      <button
        aria-label="trash-button"
        data-testid="trash-button"
        onClick={(e) => {
          e.stopPropagation();
          setOpen(true);
        }}
        className="p-1 h-auto cursor-pointer text-gray-400 hover:text-red-500"
      >
        <Trash className="w-4 h-4" />
      </button>

      <ConfirmDeleteDialog
        open={open}
        onOpenChange={setOpen}
        characterName={character.name as string}
        onConfirm={() => toggleDelete(character.id as string)}
      />
    </div>
  );
}
