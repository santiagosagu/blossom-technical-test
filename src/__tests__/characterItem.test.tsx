/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import CharacterItem from "@/components/CharacterItem";

const mockToggleFavorite = vi.fn();
const mockToggleDelete = vi.fn();
const mockSetSelectedCharacter = vi.fn();
const mockNavigate = vi.fn();

vi.mock("@/store/characterStore", () => ({
  useCharacter: () => ({
    toggleFavorite: mockToggleFavorite,
    toggleDelete: mockToggleDelete,
    setSelectedCharacter: mockSetSelectedCharacter,
  }),
}));

vi.mock("react-router", async () => {
  const actual = await vi.importActual("react-router");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

vi.mock("@/components/ui/ConfirmDialog", () => ({
  ConfirmDeleteDialog: ({
    open,
    characterName,
    onConfirm,
    onOpenChange,
  }: any) =>
    open ? (
      <div data-testid="delete-dialog">
        <p>Delete {characterName}?</p>
        <button onClick={onConfirm} data-testid="confirm">
          Yes
        </button>
        <button onClick={() => onOpenChange(false)} data-testid="cancel">
          No
        </button>
      </div>
    ) : null,
}));

const mockCharacter = {
  id: "1",
  name: "Rick Sanchez",
  species: "Human",
  image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
  isFavorite: false,
  isDeleted: false,
};

const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <BrowserRouter>{children}</BrowserRouter>
);

describe("CharacterItem - Regression Tests", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("should render without crashing and maintain basic structure", () => {
    const { container } = render(
      <TestWrapper>
        <CharacterItem character={mockCharacter} isSelected={false} />
      </TestWrapper>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render correctly when selected", () => {
    const { container } = render(
      <TestWrapper>
        <CharacterItem character={mockCharacter} isSelected={true} />
      </TestWrapper>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render correctly when character is favorite", () => {
    const favoriteCharacter = { ...mockCharacter, isFavorite: true };

    const { container } = render(
      <TestWrapper>
        <CharacterItem character={favoriteCharacter} isSelected={false} />
      </TestWrapper>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should handle core interactions correctly", () => {
    render(
      <TestWrapper>
        <CharacterItem character={mockCharacter} isSelected={false} />
      </TestWrapper>
    );

    fireEvent.click(screen.getByText("Rick Sanchez"));
    expect(mockSetSelectedCharacter).toHaveBeenCalledWith(mockCharacter);
    expect(mockNavigate).toHaveBeenCalledWith("/character-detail/1");

    fireEvent.click(screen.getAllByTestId("heart-button")[0]);
    expect(mockToggleFavorite).toHaveBeenCalledWith("1");

    fireEvent.click(screen.getAllByTestId("trash-button")[0]);
    expect(screen.getByTestId("delete-dialog")).toBeInTheDocument();
  });

  it("should contain all required elements", () => {
    render(
      <TestWrapper>
        <CharacterItem character={mockCharacter} isSelected={false} />
      </TestWrapper>
    );

    expect(screen.getByText("Rick Sanchez")).toBeInTheDocument();
    expect(screen.getByText("Human")).toBeInTheDocument();
    expect(screen.getByAltText("Rick Sanchez")).toBeInTheDocument();
    expect(screen.getAllByTestId("heart-button")[0]).toBeInTheDocument();
    expect(screen.getAllByTestId("trash-button")[0]).toBeInTheDocument();
  });

  it("should handle edge cases in character data", () => {
    const edgeCaseCharacter = {
      id: "999",
      name: "Character with very long name that should be truncated properly",
      species: "Unknown alien species with long description",
      image: "https://example.com/very-long-url-that-should-work.jpeg",
      isFavorite: true,
      isDeleted: false,
    };

    const { container } = render(
      <TestWrapper>
        <CharacterItem character={edgeCaseCharacter} isSelected={true} />
      </TestWrapper>
    );

    expect(container.firstChild).toBeInTheDocument();
    expect(
      screen.getByText(
        "Character with very long name that should be truncated properly"
      )
    ).toBeInTheDocument();
  });
});
