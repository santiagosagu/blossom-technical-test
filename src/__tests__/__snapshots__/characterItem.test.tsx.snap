// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterItem - Regression Tests > should render correctly when character is favorite 1`] = `
<div
  class="flex items-center gap-3 p-3  cursor-pointer transition-colors border-t-gray-100 border-t-1 md:rounded-lg  hover:bg-gray-100 "
>
  <div
    class="relative"
  >
    <img
      alt="Rick Sanchez"
      class="w-10 h-10 rounded-full object-cover"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
  </div>
  <div
    class="flex-1 min-w-0"
  >
    <h3
      class="font-medium text-sm truncate"
    >
      Rick Sanchez
    </h3>
    <p
      class="text-xs text-gray-500"
    >
      Human
    </p>
  </div>
  <button
    aria-label="favorite-button"
    class="p-1 h-auto cursor-pointer"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-heart w-4 h-4 fill-secondary-600 text-secondary-600"
      data-testid="heart-button"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
      />
    </svg>
  </button>
  <button
    aria-label="trash-button"
    class="p-1 h-auto cursor-pointer text-gray-400 hover:text-red-500"
    data-testid="trash-button"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-trash w-4 h-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6"
      />
      <path
        d="M3 6h18"
      />
      <path
        d="M8 6V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
      />
    </svg>
  </button>
</div>
`;

exports[`CharacterItem - Regression Tests > should render correctly when selected 1`] = `
<div
  class="flex items-center gap-3 p-3  cursor-pointer transition-colors border-t-gray-100 border-t-1 md:rounded-lg  md:bg-primary-100 "
>
  <div
    class="relative"
  >
    <img
      alt="Rick Sanchez"
      class="w-10 h-10 rounded-full object-cover"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
  </div>
  <div
    class="flex-1 min-w-0"
  >
    <h3
      class="font-medium text-sm truncate"
    >
      Rick Sanchez
    </h3>
    <p
      class="text-xs text-gray-500"
    >
      Human
    </p>
  </div>
  <button
    aria-label="favorite-button"
    class="p-1 h-auto cursor-pointer"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-heart w-4 h-4 text-gray-400"
      data-testid="heart-button"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
      />
    </svg>
  </button>
  <button
    aria-label="trash-button"
    class="p-1 h-auto cursor-pointer text-gray-400 hover:text-red-500"
    data-testid="trash-button"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-trash w-4 h-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6"
      />
      <path
        d="M3 6h18"
      />
      <path
        d="M8 6V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
      />
    </svg>
  </button>
</div>
`;

exports[`CharacterItem - Regression Tests > should render without crashing and maintain basic structure 1`] = `
<div
  class="flex items-center gap-3 p-3  cursor-pointer transition-colors border-t-gray-100 border-t-1 md:rounded-lg  hover:bg-gray-100 "
>
  <div
    class="relative"
  >
    <img
      alt="Rick Sanchez"
      class="w-10 h-10 rounded-full object-cover"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
  </div>
  <div
    class="flex-1 min-w-0"
  >
    <h3
      class="font-medium text-sm truncate"
    >
      Rick Sanchez
    </h3>
    <p
      class="text-xs text-gray-500"
    >
      Human
    </p>
  </div>
  <button
    aria-label="favorite-button"
    class="p-1 h-auto cursor-pointer"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-heart w-4 h-4 text-gray-400"
      data-testid="heart-button"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
      />
    </svg>
  </button>
  <button
    aria-label="trash-button"
    class="p-1 h-auto cursor-pointer text-gray-400 hover:text-red-500"
    data-testid="trash-button"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-trash w-4 h-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6"
      />
      <path
        d="M3 6h18"
      />
      <path
        d="M8 6V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
      />
    </svg>
  </button>
</div>
`;
